proof.go

```diff
@@ -80,90 +81,56 @@ func calcQ(e, d *bls.G1Point, y, w *bls.Fr) bls.Fr {
 	digest.Write(tmpBytes[:])
 
 	var tmp bls.Fr
-	hashToFr(&tmp, common.BytesToHash(digest.Sum(nil)))
+	hashToFr(&tmp, common.BytesToHash(digest.Sum(nil)), modulus)
 	return tmp
 }
 
-// Compute a function in eval form at one of the points in the domain
-func innerQuotients(f []bls.Fr, index int) []bls.Fr {
-	var q [nodeWidth]bls.Fr
-
-	y := f[index]
-	for i := 0; i < nodeWidth; i++ {
-		if i != index {
-			omegaIdx := (len(omegaIs) - i) % len(omegaIs)
-			invIdx := (index + nodeWidth - i) % nodeWidth
-			iMinIdx := (i - index + nodeWidth) % nodeWidth
-
-			// calculate q[i]
-			var tmp bls.Fr
-			bls.SubModFr(&tmp, &f[i], &y)
-			bls.MulModFr(&tmp, &tmp, &omegaIs[omegaIdx])
-			bls.MulModFr(&q[i], &tmp, &inverses[invIdx])
-
-			// calculate q[i]'s contribution to q[index]
-			bls.MulModFr(&tmp, &omegaIs[iMinIdx], &q[i])
-			bls.SubModFr(&tmp, &bls.ZERO, &tmp)
-			bls.AddModFr(&q[index], &q[index], &tmp)
-		}
-	}
-
-	return q[:]
+func ComputeKZGProof(tc *TreeConfig, poly []bls.Fr, z, y *bls.Fr) *bls.G1Point {
+	oq := tc.outerQuotients(poly, z, y)
+	return kzg.CommitToEvalPoly(tc.lg1, oq)
 }
 
-// Compute a function in eval form at a point outside of the domain
-func outerQuotients(f []bls.Fr, z, y *bls.Fr) []bls.Fr {
-	var q [nodeWidth]bls.Fr
-
-	for i := 0; i < nodeWidth; i++ {
-		var tmp, quo bls.Fr
-		bls.SubModFr(&tmp, &f[i], y)
-		bls.SubModFr(&quo, &omegaIs[i], z)
-		bls.DivModFr(&q[i], &tmp, &quo)
+func MakeVerkleProofOneLeaf(root VerkleNode, key []byte) (d *bls.G1Point, y *bls.Fr, sigma *bls.G1Point) {
+	var tc *TreeConfig
+	if root, ok := root.(*InternalNode); !ok {
+		panic("no tree config")
+	} else {
+		tc = root.treeConfig
 	}
 
-	return q[:]
-}
-
-func ComputeKZGProof(poly []bls.Fr, z, y *bls.Fr, lg1 []bls.G1Point) *bls.G1Point {
-	oq := outerQuotients(poly, z, y)
-	return kzg.CommitToEvalPoly(lg1, oq)
-}
-
-func MakeVerkleProofOneLeaf(root VerkleNode, key []byte, lg1 []bls.G1Point) (d *bls.G1Point, y *bls.Fr, sigma *bls.G1Point) {
 	var fis [][]bls.Fr
 	commitments, zis, yis, fis := root.GetCommitmentsAlongPath(key)
 
 	// Construct g(x)
-	r := calcR(commitments, zis, yis)
+	r := calcR(commitments, zis, yis, tc.modulus)
 
-	var g [nodeWidth]bls.Fr
+	g := make([]bls.Fr, tc.nodeWidth)
 	var powR bls.Fr
 	bls.CopyFr(&powR, &bls.ONE)
 	for index, f := range fis {
-		quotients := innerQuotients(f, index)
+		quotients := tc.innerQuotients(f, index)
 		var tmp bls.Fr
-		for i := 0; i < nodeWidth; i++ {
+		for i := 0; i < tc.nodeWidth; i++ {
 			bls.MulModFr(&tmp, &powR, &quotients[i])
 			bls.AddModFr(&g[i], &g[i], &tmp)
 		}
 
 		// rⁱ⁺¹ = r ⨯ rⁱ
 		bls.MulModFr(&powR, &powR, &r)
 	}
-	d = bls.LinCombG1(lg1, g[:])
```

Comment:
I just noticed `LinCombG1` is being used here. Wondering if the same trick for switching algorithms could help here too

Reply:
most likely :+1:

Reply:
Actually, let me take this back: `g` is calculated on the evaluation basis, so it's unlikely that any coefficient will be zero.